// cdk-iot-core-certificates-v3/lib/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getParamStorePath = exports.ThingWithCert = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const node_path_1 = require("node:path");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const cfn = require("aws-cdk-lib/aws-cloudformation");
const iam = require("aws-cdk-lib/aws-iam");
const lambda = require("aws-cdk-lib/aws-lambda");
const nodejs = require("aws-cdk-lib/aws-lambda-nodejs");
const logs = require("aws-cdk-lib/aws-logs");
const s3deploy = require("aws-cdk-lib/aws-s3-deployment");
const ssm = require("aws-cdk-lib/aws-ssm");
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
const constructs_1 = require("constructs");
/**
 * A CDK Construct for creating an AWS IoT Core thing with a certificate
 */
class ThingWithCert extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { thingName, saveToParamStore, paramPrefix, saveFileBucket } = props;
        const role = new iam.Role(this, 'LambdaExecutionRole', {
            assumedBy: new iam.CompositePrincipal(new iam.ServicePrincipal('lambda.amazonaws.com')),
        });
        role.addToPolicy(new iam.PolicyStatement({
            resources: ['arn:aws:logs:*:*:*'],
            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
        }));
        role.addToPolicy(new iam.PolicyStatement({
            resources: ['*'],
            actions: ['iot:*'],
        }));
        const onEventHandler = new nodejs.NodejsFunction(this, 'lambdaFunction', {
            entry: (0, node_path_1.join)(__dirname, 'lambda', 'index.js'),
            handler: 'handler',
            timeout: aws_cdk_lib_1.Duration.seconds(10),
            role,
            runtime: lambda.Runtime.NODEJS_LATEST,
            logRetention: logs.RetentionDays.ONE_DAY,
        });
        const { serviceToken } = new custom_resources_1.Provider(this, 'lambdaProvider', {
            onEventHandler,
        });
        const lambdaCustomResource = new cfn.CfnCustomResource(this, 'lambdaCustomResource', {
            serviceToken,
        });
        lambdaCustomResource.addPropertyOverride('ThingName', thingName);
        const paramStorePath = (0, exports.getParamStorePath)(thingName, paramPrefix);
        if (saveToParamStore) {
            new ssm.CfnParameter(this, 'paramStoreCertPem', {
                type: 'String',
                value: lambdaCustomResource.getAtt('certPem').toString(),
                name: `${paramStorePath}/certPem`,
            });
            new ssm.CfnParameter(this, 'paramStorePrivKey', {
                type: 'String',
                value: lambdaCustomResource.getAtt('privKey').toString(),
                name: `${paramStorePath}/privKey`,
            });
        }
        this.thingArn = lambdaCustomResource.getAtt('thingArn').toString();
        this.certId = lambdaCustomResource.getAtt('certId').toString();
        this.certPem = lambdaCustomResource.getAtt('certPem').toString();
        this.privKey = lambdaCustomResource.getAtt('privKey').toString();
        if (saveFileBucket) {
            new s3deploy.BucketDeployment(this, 'SaveFile', {
                sources: [
                    s3deploy.Source.data(`${thingName}/${thingName}.cert.pem`, this.certPem),
                    s3deploy.Source.data(`${thingName}/${thingName}.private.key`, this.privKey),
                ],
                destinationBucket: saveFileBucket,
            });
        }
    }
}
exports.ThingWithCert = ThingWithCert;
_a = JSII_RTTI_SYMBOL_1;
ThingWithCert[_a] = { fqn: "cdk-iot-core-certificates-v3.ThingWithCert", version: "0.0.14" };
const getParamStorePath = (thingName, paramPrefix) => {
    if (thingName.charAt(0) === '/') {
        throw new Error("thingName cannot start with '/'");
    }
    if (paramPrefix && paramPrefix.charAt(0) === '/') {
        throw new Error("paramPrefix cannot start with '/'");
    }
    return paramPrefix ? `/${paramPrefix}/${thingName}` : `/${thingName}`;
};
exports.getParamStorePath = getParamStorePath;
