// cdk-iot-core-certificates-v3/lib/lambda/adapters/iot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iotAdaptor = void 0;
const iot_1 = require("../util/iot");
const iotAdaptor = (iot) => {
    return {
        createThing: async (thingRequest) => {
            return iot.createThing(thingRequest);
        },
        createKeysAndCertificates: async () => {
            return iot
                .createKeysAndCertificate({
                setAsActive: true,
            });
        },
        createPolicy: async (thingName) => {
            return iot
                .createPolicy({
                policyName: thingName,
                policyDocument: policyDoc,
            });
        },
        attachPrincipalPolicy: async (props) => {
            await iot.attachPrincipalPolicy(props);
        },
        attachThingPrincipal: async (props) => {
            return iot.attachThingPrincipal(props);
        },
        listThingPrincipals: async (thingName) => {
            return iot
                .listThingPrincipals({
                thingName: thingName,
            });
        },
        detachPrincipalPolicy: async (props) => {
            await iot.detachPrincipalPolicy(props);
        },
        detachThingPrincipal: async (props) => {
            return iot.detachThingPrincipal(props);
        },
        updateCertificateToInactive: async (certArn) => {
            await iot
                .updateCertificate({
                certificateId: (0, iot_1.getCertIdFromARN)(certArn),
                newStatus: 'INACTIVE',
            });
        },
        deleteCertificate: async (certArn) => {
            await iot
                .deleteCertificate({
                certificateId: (0, iot_1.getCertIdFromARN)(certArn),
            });
        },
        deletePolicy: async (policyName) => {
            await iot
                .deletePolicy({
                policyName: policyName,
            });
        },
        deleteThing: async (thingName) => {
            await iot
                .deleteThing({
                thingName: thingName,
            });
        },
    };
};
exports.iotAdaptor = iotAdaptor;
const policyDoc = `{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:Publish",
        "iot:Subscribe",
        "iot:Connect",
        "iot:Receive"
      ],
      "Resource": [
        "*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "iot:GetThingShadow",
        "iot:UpdateThingShadow",
        "iot:DeleteThingShadow"
      ],
      "Resource": [
        "*"
      ]
    }
  ]
}`;
